### PR TITLE
feat(expression): support map data type

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

- Scalar support map value
- Column support map column
- Add map function `map(array(keys), array(values))`

part of #10062

### CODE REVIEW

**Comment by BohuTANG** [comment]
compact test all failed , cc @drmingdrmer

**Comment by drmingdrmer** [comment]
> compact test all failed , cc @drmingdrmer

It looks the following failure is the first time it fails:
https://github.com/datafuselabs/databend/pull/10144#issuecomment-1438307226

Let's check what's changed in that PR

**Comment by BohuTANG** [comment]
@mergify update

**Comment by drmingdrmer** [comment]
meta-compat test is fixed in:
- #10153

**Comment by BohuTANG** [comment]
```
[SQL] SELECT name, value, default, level from system.settings where name in ('sql_dialect', 'timezone', 'wait_for_async_insert_timeout')
[Diff] (-expected|+actual)
    sql_dialect PostgreSQL PostgreSQL SESSION
-   timezone                       UTC         UTC         SESSION
-   wait_for_async_insert_timeout  100         100         SESSION
+   timezone Asia/Shanghai UTC SESSION
+   wait_for_async_insert_timeout 20000 100 SESSION
at tests/sqllogictests/suites/base/05_ddl/05_0026_ddl_unset_settings:42

Error: SelfError("sqllogictest failed")
Error: Process completed with exit code 1.
```

File: src/query/expression/src/types/map.rs [hunk 1]
```diff
@@ -298,110 +297,110 @@ unsafe impl<'a, K: ValueType, V: ValueType> TrustedLen for KvIterator<'a, K, V>
...
     }
 
     fn try_downcast_scalar<'a>(scalar: &'a ScalarRef) -> Option<Self::ScalarRef<'a>> {
-        <MapInternal<T> as ValueType>::try_downcast_scalar(scalar)
+        <MapInternal<K, V> as ValueType>::try_downcast_scalar(scalar)
     }
 
     fn try_downcast_column<'a>(col: &'a Column) -> Option<Self::Column> {
-        <MapInternal<T> as ValueType>::try_downcast_column(col)
+        <MapInternal<K, V> as ValueType>::try_downcast_column(col)
```
Comment (andylokandy) [inline src/query/expression/src/types/map.rs:330]: Should downcast `Column::Map`, but not `Column::Array`

File: src/query/expression/src/types/variant.rs [hunk 2]
```diff
@@ -197,6 +197,33 @@ pub fn cast_scalar_to_variant(scalar: ScalarRef, tz: TzLUT, buf: &mut Vec<u8>) {
             common_jsonb::build_array(items.iter(), buf).expect("failed to build jsonb array");
             return;
         }
+        ScalarRef::Map(col) => {
+            let kvs = col
+                .iter()
+                .map(|s| match s {
+                    ScalarRef::Tuple(t) => {
+                        let key = match t[0] {
```
Comment (andylokandy) [inline src/query/expression/src/types/variant.rs:205]: It's better to downcast via `MapType<AnyType, AnyType>::try_downcast_column`

File: src/query/functions/src/scalars/map.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+use common_expression::ValueRef;
+use common_hashtable::StackHashSet;
+use siphasher::sip128::Hasher128;
+use siphasher::sip128::SipHasher24;
+
+pub fn register(registry: &mut FunctionRegistry) {
+    registry.register_function_factory("map", |_, args_type| {
+        if args_type.len() != 2 {
+            return None;
+        }
```
Comment (andylokandy) [inline src/query/functions/src/scalars/map.rs:41]: it's better to use `registry.register_passthrough_nullable::<ArrayType<GenericType<0>>, ArrayType<GenericType<1>>, MapType<GenericType<0>, GenericType<1>>>` and then check the key type via `ctx.generics[0]`

File: src/query/functions/tests/it/scalars/map.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    test_create(file);
+}
+
+fn test_create(file: &mut impl Write) {
+    run_ast(file, "map([1,2], ['a','b'])", &[]);
+    run_ast(file, "map(['k1','k2','k3'], ['v1','v2','v3'])", &[]);
+
+    run_ast(file, "map(1, 'a')", &[]);
+    run_ast(file, "map(['k1','k2'], ['v1','v2','v3'])", &[]);
+    run_ast(file, "map(['k1','k1'], ['v1','v2'])", &[]);
```
Comment (andylokandy) [inline src/query/functions/tests/it/scalars/map.rs:35]: may need test cases for columns.

File: src/meta/proto-conv/src/schema_from_to_protobuf_impl.rs [hunk 5]
```diff
@@ -223,6 +223,7 @@ impl FromToProto for ex::TableDataType {
         let x = match self {
             TableDataType::Null => new_pb_dt24(Dt24::NullT(pb::Empty {})),
             TableDataType::EmptyArray => new_pb_dt24(Dt24::EmptyArrayT(pb::Empty {})),
+            TableDataType::EmptyMap => todo!(),
```
Comment (drmingdrmer) [inline src/meta/proto-conv/src/schema_from_to_protobuf_impl.rs:226]: Will this `todo!()` be removed?

Comment (drmingdrmer) [reply src/meta/proto-conv/src/schema_from_to_protobuf_impl.rs:226]: Yes, and if proto-conv is modified, the a new proto version should be added, and the corresponding test should be added:

https://github.com/datafuselabs/databend/blob/c3bcdd27b631107df0d1e464b5d384fede8a3d53/src/meta/proto-conv/src/util.rs#L89-L94

File: src/query/expression/src/types.rs [hunk 6]
```diff
@@ -285,7 +288,7 @@ impl DataType {
             DataType::Date => Scalar::Date(0),
             DataType::Nullable(_) => Scalar::Null,
             DataType::Array(_) => Scalar::EmptyArray,
-            DataType::Map(_) => todo!(),
+            DataType::Map(_) => Scalar::EmptyMap,
```
Comment (andylokandy) [inline src/query/expression/src/types.rs:291]: The type is not matched.

File: src/query/expression/src/types/variant.rs [hunk 7]
```diff
@@ -197,6 +197,33 @@ pub fn cast_scalar_to_variant(scalar: ScalarRef, tz: TzLUT, buf: &mut Vec<u8>) {
...
+                            _ => unreachable!(),
+                        };
+                        let mut val = vec![];
+                        cast_scalar_to_variant(t[1].clone(), tz, &mut val);
+                        (key, val)
+                    }
+                    _ => unreachable!(),
+                })
+                .collect::<Vec<_>>();
+            common_jsonb::build_object(kvs.iter().map(|(k, v)| (k, &v[..])), buf)
```
[the remark this answers is not in the document]
Comment (b41sh) [reply src/query/expression/src/types/variant.rs:224]: we can't borrow `v` here, which will cause the following error.
`returns a value referencing data owned by the current function`

File: src/query/expression/src/types/map.rs (hunk 1 above)
Comment (b41sh) [reply src/query/expression/src/types/map.rs:333]: `Column::Map` is also an `ArrayColumn`.

File: src/meta/proto-conv/tests/it/main.rs [hunk 8]
```diff
@@ -34,3 +34,5 @@ mod v024_table_meta;
 mod v025_user_stage;
 mod v026_schema;
 mod v027_schema;
+mod v028_schema;
+mod v029_schema;
```
Comment (drmingdrmer) [inline src/meta/proto-conv/tests/it/main.rs:38]: Is there a `v028_schema` file added?

File: src/meta/proto-conv/tests/it/v029_schema.rs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            "decimal256",
+            TableDataType::Decimal(DecimalDataType::Decimal256(DecimalSize {
+                precision: 46,
+                scale: 6,
+            })),
+        ),
+    ];
+    let want = || TableSchema::new(fields.clone());
+    common::test_load_old(func_name!(), schema_v29.as_slice(), 29, want())?;
+    common::test_pb_from_to(func_name!(), want())?;
```
Comment (drmingdrmer) [inline src/meta/proto-conv/tests/it/v029_schema.rs:81]: This test should include the change introduced in this PR

File: src/meta/proto-conv/tests/it/main.rs (hunk 8 above)
Comment (b41sh) [reply src/meta/proto-conv/tests/it/main.rs:38]: yes, 0028 is `2023-02-13: Add: user.proto/UserDefinedFileFormat`

File: src/meta/protos/proto/datatype.proto [hunk 10]
```diff
@@ -61,6 +61,7 @@ message DataType {
     Tuple    tuple_t       = 41;
     Empty    variant_t     = 42;
     Decimal  decimal_t     = 43;
+    Empty    empty_map_t   = 44;
```
Comment (sundy-li) [inline src/meta/protos/proto/datatype.proto:64]: Two Empty

File: src/meta/proto-conv/tests/it/v028_schema.rs [hunk 11]
```diff
@@ -34,19 +34,19 @@ use crate::common;
...
+        28, 168, 6, 24, 18, 17, 154, 2, 8, 66, 0, 160, 6, 28, 168, 6, 24, 160, 6, 28, 168, 6, 24,
+        160, 6, 28, 168, 6, 24, 160, 6, 28, 168, 6, 24, 32, 1, 160, 6, 28, 168, 6, 24, 10, 30, 10,
+        1, 99, 26, 17, 154, 2, 8, 34, 0, 160, 6, 28, 168, 6, 24, 160, 6, 28, 168, 6, 24, 32, 4,
+        160, 6, 28, 168, 6, 24, 10, 49, 10, 10, 100, 101, 99, 105, 109, 97, 108, 49, 50, 56, 26,
+        27, 218, 2, 18, 10, 10, 8, 18, 16, 3, 160, 6, 28, 168, 6, 24, 160, 6, 28, 168, 6, 24, 160,
+        6, 28, 168, 6, 24, 32, 5, 160, 6, 28, 168, 6, 24, 10, 49, 10, 10, 100, 101, 99, 105, 109,
+        97, 108, 50, 53, 54, 26, 27, 218, 2, 18, 18, 10, 8, 46, 16, 6, 160, 6, 28, 168, 6, 24, 160,
+        6, 28, 168, 6, 24, 160, 6, 28, 168, 6, 24, 32, 6, 160, 6, 28, 168, 6, 24, 24, 7, 160, 6,
+        28, 168, 6, 24,
     ];
```
Comment (drmingdrmer) [inline src/meta/proto-conv/tests/it/v028_schema.rs:50]: Hmmm... why are the bytes changed? These data is built with older version and is meant to assert compatibility. Therefore once it is created it should not be updated.

File: src/meta/proto-conv/tests/it/v029_schema.rs [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+            TableDataType::Map(Box::new(TableDataType::Tuple {
+                fields_name: vec!["key".to_string(), "val".to_string()],
+                fields_type: vec![
+                    TableDataType::String,
+                    TableDataType::Number(NumberDataType::Int64),
+                ],
+            })),
+        ),
+    ];
+    let want = || TableSchema::new(fields.clone());
```
Comment (drmingdrmer) [inline src/meta/proto-conv/tests/it/v029_schema.rs:95]: The other `Tablefield`s except those that contain EmptyMap are not necessary for this test IMHO.

File: src/meta/proto-conv/tests/it/v028_schema.rs (hunk 11 above)
Comment (b41sh) [reply src/meta/proto-conv/tests/it/v028_schema.rs:50]: `0028` is not added to `main.rs`, the previous bytes are copied from `0027`, so it's not correct.

File: src/meta/proto-conv/tests/it/v029_schema.rs (hunk 12 above)
Comment (b41sh) [reply src/meta/proto-conv/tests/it/v029_schema.rs:95]: ok, I will remove EmptyMap.

**Comment by lichuang** [comment]
please add map type test in function `schema.rs::create_test_complex_schema`, test/schema.rs and test/column_node.rs use this function to generate test schema, and please add this type test case in there test cases.

**Comment by b41sh** [comment]
> please add map type test in function `schema.rs::create_test_complex_schema`, test/schema.rs and test/column_node.rs use this function to generate test schema, and please add this type test case in there test cases.

@lichuang Thanks for the heads up, but modifying the `create_test_complex_schema` will involve generating `leaf_id` for the `Map` data type and will require a lot of code changes. In order to keep this PR simple and merge it ASAP, I will modify `create_test_complex_schema` and generate the `leaf_id` in the next PR.

File: src/query/expression/src/evaluator.rs [hunk 13]
```diff
@@ -312,7 +312,40 @@ impl<'a> Evaluator<'a> {
...
+                }
+                Value::Column(Column::Map(col)) => {
+                    let new_col = self
+                        .run_cast(span, inner_src_ty, inner_dest_ty, Value::Column(col.values))?
+                        .into_column()
+                        .unwrap();
+                    Ok(Value::Column(Column::Map(Box::new(ArrayColumn {
+                        values: new_col,
+                        offsets: col.offsets,
+                    }))))
```
Comment (andylokandy) [inline src/query/expression/src/evaluator.rs:345]: does it work?
```suggestion
                    Ok(Value::Column(Column::Map(Box::new(new_col))))
```

File: src/query/expression/src/type_check.rs [hunk 14]
```diff
@@ -394,13 +405,29 @@ pub fn unify(
...
+                    .iter()
+                    .zip(dest_tys)
+                    .map(|(src_ty, dest_ty)| unify(src_ty, dest_ty, auto_cast_rules))
+                    .collect::<Result<Vec<_>>>()?;
+                let subst = substs
+                    .into_iter()
+                    .try_reduce(|subst1, subst2| subst1.merge(subst2, auto_cast_rules))?
+                    .unwrap_or_else(Substitution::empty);
+                Ok(subst)
+            }
```
Comment (andylokandy) [inline src/query/expression/src/type_check.rs:428]:
```suggestion
            (DataType::Tuple(_), DataType::Tuple(_)) => {
                unify(src_tys, dest_tys, auto_cast_rules)
            }
```

File: src/query/expression/src/type_check.rs [hunk 15]
```diff
@@ -438,13 +465,21 @@ pub fn can_auto_cast_to(
...
         }
         (src_ty, DataType::Nullable(dest_ty)) => can_auto_cast_to(src_ty, dest_ty, auto_cast_rules),
         (DataType::Array(src_ty), DataType::Array(dest_ty)) => {
             can_auto_cast_to(src_ty, dest_ty, auto_cast_rules)
         }
+        (DataType::Map(box src_ty), DataType::Map(box dest_ty)) => match (src_ty, dest_ty) {
+            (DataType::Tuple(src_tys), DataType::Tuple(dest_tys)) => src_tys
+                .iter()
+                .zip(dest_tys)
+                .all(|(src_ty, dest_ty)| can_auto_cast_to(src_ty, dest_ty, auto_cast_rules)),
```
Comment (andylokandy) [inline src/query/expression/src/type_check.rs:480]:
```suggestion
            (DataType::Tuple(_), DataType::Tuple(_)) => can_auto_cast_to(src_ty, dest_ty, auto_cast_rules),
```

File: src/query/expression/src/type_check.rs [hunk 16]
```diff
@@ -479,6 +514,20 @@ pub fn common_super_type(
...
+        (DataType::Map(box ty1), DataType::Map(box ty2)) => match (ty1, ty2) {
+            (DataType::Tuple(tys1), DataType::Tuple(tys2)) => {
+                let tys = tys1
+                    .into_iter()
+                    .zip(tys2)
+                    .map(|(ty1, ty2)| common_super_type(ty1, ty2, auto_cast_rules))
+                    .collect::<Option<Vec<_>>>()?;
+                let inner_ty = DataType::Tuple(tys);
+                Some(DataType::Map(Box::new(inner_ty)))
+            }
```
Comment (andylokandy) [inline src/query/expression/src/type_check.rs:528]:
```suggestion
            (DataType::Tuple(_), DataType::Tuple(_)) =>  common_super_type(ty1, ty2, auto_cast_rules)
```

File: src/query/expression/src/values.rs [hunk 17]
```diff
@@ -942,6 +1002,9 @@ impl Column {
             ArrowDataType::Extension(name, _, _) if name == "EmptyArray" => Column::EmptyArray {
                 len: arrow_col.len(),
             },
+            ArrowDataType::Extension(name, _, _) if name == "EmptyMap" => Column::EmptyMap {
```
Comment (andylokandy) [inline src/query/expression/src/values.rs:1005]:
```suggestion
            ArrowDataType::Extension(name, _, _) if name == ARROW_EXT_TYPE_EMPTY_MAP => Column::EmptyMap {
```

File: src/query/expression/src/values.rs [hunk 18]
```diff
@@ -942,6 +1002,9 @@ impl Column {
             ArrowDataType::Extension(name, _, _) if name == "EmptyArray" => Column::EmptyArray {
```
Comment (andylokandy) [inline src/query/expression/src/values.rs:1002]: ditto

File: src/query/expression/src/types/map.rs (hunk 1 above)
Comment (andylokandy) [reply src/query/expression/src/types/map.rs:333]: MapInternal will downcast `Column::Array` instead of `Column::Map`

Comment (b41sh) [reply src/query/expression/src/types/map.rs:333]: ok

File: src/meta/proto-conv/tests/it/v029_schema.rs (hunk 12 above)
Comment (drmingdrmer) [reply src/meta/proto-conv/tests/it/v029_schema.rs]: Sorry, I mean, you only have to test EmptyMap. Since it is the only change in this PR...

Comment (b41sh) [reply src/meta/proto-conv/tests/it/v029_schema.rs]: ok

**Review by andylokandy — APPROVED** [review]

**Pull request merged**